crates/worker/src/partition/mod.rs

```diff
@@ -355,21 +381,29 @@ where
             .map_ok(|entry| {
                 trace!(?entry, "Read entry");
                 let lsn = entry.sequence_number();
-                let Some(envelope) = entry.try_decode_arc::<Envelope>() else {
-                    // trim-gap
-                    unimplemented!("Handling trim gap is currently not supported")
-                };
-                anyhow::Ok((lsn, envelope?))
+                if entry.is_data_record() {
+                    entry
+                        .try_decode_arc::<Envelope>()
+                        .map(|envelope| anyhow::Ok(Record::Envelope(lsn, envelope?)))
+                        .expect("data record is present")
+                } else {
+                    anyhow::Ok(Record::TrimGap(
+                        entry
+                            .trim_gap_to_sequence_number()
+                            .expect("trim gap has to-LSN"),
+                    ))
```

Comment:
Would it make sense to stop the read stream at the first gap and return Err instead?

Reply:
I understand this to mean the `map_ok` function translates a trim-gap into an `Err(TrimGap {..})` instead? Maybe! The way we use `anyhow::Result` pervasively makes this a deeper change than I wanted to tackle right away; but it also makes more sense to treat trim gaps as just another record in the stream, with errors reserved for actual failure conditions.

Zooming out a bit, modeling the Partition Processor overall outcome as `Result<Canceled | StoppedAtTrimGap, ProcessingError>` seems accurate: the Ok / left path is an expected if rare reason to halt; the Err / right path is an exceptional failure condition.

If you have a few minutes, I'd love to hear more about how you'd solve this? I'm certain I am also missing some subtlety around properly consuming the log stream!

Reply:
We discussed offline and agreed that it's best to represent this case as an error case.